In [36]:
import pandas as pd
import numpy as np
#import matplotlib.pyplot as plt
import statsmodels.api as sm
from sklearn.model_selection import train_test_split # 학습용/평가용 데이터 나누기
from sklearn.preprocessing import StandardScaler #전처리 - 표준화 하기 

In [3]:
df = pd.read_csv('c:/bd092_data/BankPersonalLoan.csv')
df.head()

ID  Age  Experience  Income  ZIP Code  Family  CCAvg  Education  Mortgage  \
0   1   25           1      49     91107       4    1.6          1         0   
1   2   45          19      34     90089       3    1.5          1         0   
2   3   39          15      11     94720       1    1.0          1         0   
3   4   35           9     100     94112       1    2.7          2         0   
4   5   35           8      45     91330       4    1.0          2         0   

   Personal Loan  Securities Account  CD Account  Online  CreditCard  
0              0                   1           0       0           0  
1              0                   1           0       0           0  
2              0                   0           0       0           0  
3              0                   0           0       0           0  
4              0                   0           0       0           1

In [6]:
df.shape

(5000, 14)

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 14 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   ID                  5000 non-null   int64  
 1   Age                 5000 non-null   int64  
 2   Experience          5000 non-null   int64  
 3   Income              5000 non-null   int64  
 4   ZIP Code            5000 non-null   int64  
 5   Family              5000 non-null   int64  
 6   CCAvg               5000 non-null   float64
 7   Education           5000 non-null   int64  
 8   Mortgage            5000 non-null   int64  
 9   Personal Loan       5000 non-null   int64  
 10  Securities Account  5000 non-null   int64  
 11  CD Account          5000 non-null   int64  
 12  Online              5000 non-null   int64  
 13  CreditCard          5000 non-null   int64  
dtypes: float64(1), int64(13)
memory usage: 547.0 KB


In [8]:
df.describe()

ID          Age   Experience       Income      ZIP Code  \
count  5000.000000  5000.000000  5000.000000  5000.000000   5000.000000   
mean   2500.500000    45.338400    20.104600    73.774200  93152.503000   
std    1443.520003    11.463166    11.467954    46.033729   2121.852197   
min       1.000000    23.000000    -3.000000     8.000000   9307.000000   
25%    1250.750000    35.000000    10.000000    39.000000  91911.000000   
50%    2500.500000    45.000000    20.000000    64.000000  93437.000000   
75%    3750.250000    55.000000    30.000000    98.000000  94608.000000   
max    5000.000000    67.000000    43.000000   224.000000  96651.000000   

            Family        CCAvg    Education     Mortgage  Personal Loan  \
count  5000.000000  5000.000000  5000.000000  5000.000000    5000.000000   
mean      2.396400     1.937938     1.881000    56.498800       0.096000   
std       1.147663     1.747659     0.839869   101.713802       0.294621   
min       1.000000     0.000000     1.000000     0.000000       0.000000   
25%       1.000000     0.700000     1.000000     0.000000       0.000000   
50%       2.000000     1.500000     2.000000     0.000000       0.000000   
75%       3.000000     2.500000     3.000000   101.000000       0.000000   
max       4.000000    10.000000     3.000000   635.000000       1.000000   

       Securities Account  CD Account       Online   CreditCard  
count         5000.000000  5000.00000  5000.000000  5000.000000  
mean             0.104400     0.06040     0.596800     0.294000  
std              0.305809     0.23825     0.490589     0.455637  
min              0.000000     0.00000     0.000000     0.000000  
25%              0.000000     0.00000     0.000000     0.000000  
50%              0.000000     0.00000     1.000000     0.000000  
75%              0.000000     0.00000     1.000000     1.000000  
max              1.000000     1.00000     1.000000     1.000000

In [12]:
df.head()

Age  Income  Family  CCAvg  Education  Mortgage  Personal Loan  \
0   25      49       4    1.6          1         0              0   
1   45      34       3    1.5          1         0              0   
2   39      11       1    1.0          1         0              0   
3   35     100       1    2.7          2         0              0   
4   35      45       4    1.0          2         0              0   

   Securities Account  CD Account  Online  CreditCard  
0                   1           0       0           0  
1                   1           0       0           0  
2                   0           0       0           0  
3                   0           0       0           0  
4                   0           0       0           1

In [23]:
df = pd.read_csv('c:/bd092_data/BankPersonalLoan.csv')
df.drop(['ID', 'Experience', 'ZIP Code'], axis=1, inplace=True)

In [26]:
# x,y 종속변수와 독립변수 분리하기 
x = df.drop('Personal Loan', axis=1)
y = df['Personal Loan']

In [66]:
#학습용 / 평가용 70% : 30%으로 나누기
x_train, x_test, y_train, y_test = train_test_split(x,y, test_size = 0.3, random_state=1, stratify=y)
print(x_train.shape, x_test.shape, y_train.shape, y_test.shape)

(3500, 10) (1500, 10) (3500,) (1500,)


In [67]:
#표준화 작업
stdsc = StandardScaler() #함수의 활성화 <변수에 담아서 더 간단하게 쓰기>
#standard scaler 함수 사용하기 
x_train.iloc[:, :] = stdsc.fit_transform(x_train.iloc[:, :])
x_test.iloc[:, :] = stdsc.transform(x_test.iloc[:, :])

C:\Users\Ki beom kim\anaconda3\lib\site-packages\pandas\core\indexing.py:966: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self.obj[item] = s
C:\Users\Ki beom kim\anaconda3\lib\site-packages\pandas\core\indexing.py:966: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self.obj[item] = s


In [68]:
# 이제 다시 모델 만들기 
log_model = sm.Logit(y_train, x_train)
result = log_model.fit()

Optimization terminated successfully.
         Current function value: 0.617262
         Iterations 6


In [70]:
result.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                           Logit Regression Results                           
==============================================================================
Dep. Variable:          Personal Loan   No. Observations:                 3500
Model:                          Logit   Df Residuals:                     3490
Method:                           MLE   Df Model:                            9
Date:                Tue, 27 Oct 2020   Pseudo R-squ.:                 -0.9521
Time:                        23:18:56   Log-Likelihood:                -2160.4
converged:                       True   LL-Null:                       -1106.7
Covariance Type:            nonrobust   LLR p-value:                     1.000
======================================================================================
                         coef    std err          z      P>|z|      [0.025      0.975]
--------------------------------------------------------------------------------------
Age                    0.0348      0.037      0.950      0.342      -0.037       0.107
Income                 0.7018      0.053     13.353      0.000       0.599       0.805
Family                 0.1863      0.038      4.962      0.000       0.113       0.260
CCAvg                  0.0630      0.049      1.275      0.202      -0.034       0.160
Education              0.3354      0.038      8.781      0.000       0.261       0.410
Mortgage               0.0474      0.040      1.200      0.230      -0.030       0.125
Securities Account    -0.1094      0.040     -2.706      0.007      -0.189      -0.030
CD Account             0.4516      0.053      8.488      0.000       0.347       0.556
Online                -0.0535      0.037     -1.454      0.146      -0.126       0.019
CreditCard            -0.0930      0.038     -2.431      0.015      -0.168      -0.018
======================================================================================
"""

In [74]:
result.params.sort_values(ascending=False)

Income                0.701833
CD Account            0.451610
Education             0.335415
Family                0.186348
CCAvg                 0.063020
Mortgage              0.047440
Age                   0.034847
Online               -0.053486
CreditCard           -0.092989
Securities Account   -0.109390
dtype: float64

In [72]:
np.exp(result.params).sort_values(ascending=False)

Income                2.017448
CD Account            1.570839
Education             1.398521
Family                1.204842
CCAvg                 1.065048
Mortgage              1.048584
Age                   1.035461
Online                0.947919
CreditCard            0.911203
Securities Account    0.896381
dtype: float64

# 해석 & 결론
#income (X값)이 증가할수록, 해당 고객이 대출성공 그룹에 '분류' 될 확률은 101.7% 증가한다. 고 해석. 
#or income 증가할수록, 해당 고객이 대출성공 그룹에 '분류'될 확률은 2.017배 증가한다. 고 해석. 
#(2.017-1 *100)%

#INCOME이 높은 사람은 대출성공에 분류될 가능성이 높은 '잠재고객'이다. 
#income 높은 잠재고객에게 프로모션하면 대출성공에 분류되는 성공률을 높일 수 있기 때문에 이 사람에게 프로모션 해야 한다. 
